File: node/function/src/db/CosmosDBDatabase.js
const { CosmosClient } = require('@azure/cosmos');
const DatabaseInterface = require('./DatabaseInterface');

class CosmosDBDatabase extends DatabaseInterface {
  constructor(connectionString, databaseName = 'urlshortener', containerName = 'urls') {
    super();
    
    // Parse connection string or use individual parameters for backward compatibility
    if (connectionString.includes('AccountEndpoint=')) {
      // Connection string format: "AccountEndpoint=https://your-account.documents.azure.com:443/;AccountKey=your-key;Database=urlshortener;"
      const parsed = this.parseConnectionString(connectionString);
      this.client = new CosmosClient({ endpoint: parsed.endpoint, key: parsed.key });
      this.databaseName = parsed.database || databaseName;
    } else {
      // Legacy format - assume first parameter is endpoint, second is key
      this.client = new CosmosClient({ endpoint: connectionString, key: databaseName });
      this.databaseName = containerName || 'urlshortener';
    }
    
    this.containerName = containerName || 'urls';
    this.database = null;
    this.container = null;
  }

  parseConnectionString(connectionString) {
    const parts = connectionString.split(';').filter(part => part.trim());
    const config = {};
    
    for (const part of parts) {
      const [key, value] = part.split('=');
      if (key && value) {
        switch (key.trim()) {
          case 'AccountEndpoint':
            config.endpoint = value.trim();
            break;
          case 'AccountKey':
            config.key = value.trim();
            break;
          case 'Database':
            config.database = value.trim();
            break;
        }
      }
    }
    
    if (!config.endpoint || !config.key) {
      throw new Error('Connection string must contain AccountEndpoint and AccountKey');
    }
    
    return config;
  }

  async initialize() {
    try {
      // Create database if it doesn't exist
      const { database } = await this.client.databases.createIfNotExists({
        id: this.databaseName
      });
      this.database = database;

      // Create container if it doesn't exist
      const { container } = await this.database.containers.createIfNotExists({
        id: this.containerName,
        partitionKey: { paths: ['/shortCode'] }
      });
      this.container = container;
    } catch (err) {
      throw new Error(`Cosmos DB initialization failed: ${err.message}`);
    }
  }

  async createUrl(shortCode, originalUrl) {
    try {
      const item = {
        id: shortCode, // Use shortCode as the document ID
        shortCode: shortCode,
        originalUrl: originalUrl,
        createdAt: new Date().toISOString()
      };

      const { resource } = await this.container.items.create(item);
      return { shortCode, originalUrl, id: resource.id };
    } catch (err) {
      throw new Error(`Failed to create URL: ${err.message}`);
    }
  }

  async getUrl(shortCode) {
    try {
      const { resource } = await this.container.item(shortCode, shortCode).read();
      if (resource) {
        return {
          shortCode: resource.shortCode,
          originalUrl: resource.originalUrl,
          createdAt: resource.createdAt
        };
      }
      return null;
    } catch (err) {
      if (err.code === 404) {
        return null;
      }
      throw new Error(`Failed to get URL: ${err.message}`);
    }
  }

  async getAllUrls() {
    try {
      const { resources } = await this.container.items
        .query('SELECT * FROM c ORDER BY c.createdAt DESC')
        .fetchAll();
      
      return resources.map(item => ({
        shortCode: item.shortCode,
        originalUrl: item.originalUrl,
        createdAt: item.createdAt
      }));
    } catch (err) {
      throw new Error(`Failed to get all URLs: ${err.message}`);
    }
  }

  async urlExists(shortCode) {
    const url = await this.getUrl(shortCode);
    return !!url;
  }

  async close() {
    // Cosmos DB client doesn't need explicit closing
  }
}

module.exports = CosmosDBDatabase;

File: dotnet/UrlShortener.Core/Database/CosmosDbDatabase.cs
using Microsoft.Azure.Cosmos;
using UrlShortener.Core.Interfaces;
using UrlShortener.Core.Models;
using System.Net;

namespace UrlShortener.Core.Database;

public class CosmosDbDatabase : IUrlDatabase
{
    private readonly string _databaseName;
    private readonly string _containerName;
    private CosmosClient? _cosmosClient;
    private Container? _container;

    public CosmosDbDatabase(string connectionString, string databaseName = "urlshortener", string containerName = "urls")
    {
        if (connectionString.Contains("AccountEndpoint="))
        {
            var parsed = ParseConnectionString(connectionString);
            _cosmosClient = new CosmosClient(parsed.endpoint, parsed.key);
            _databaseName = parsed.database ?? databaseName;
        }
        else
        {
            // Legacy format - assume first parameter is endpoint, second is key
            throw new ArgumentException("Connection string format not supported in this constructor. Use endpoint and key parameters.");
        }
        _containerName = containerName;
    }

    public CosmosDbDatabase(string endpoint, string key, string databaseName = "urlshortener", string containerName = "urls")
    {
        _cosmosClient = new CosmosClient(endpoint, key);
        _databaseName = databaseName;
        _containerName = containerName;
    }

    public async Task InitializeAsync()
    {
        if (_cosmosClient == null)
            throw new InvalidOperationException("Cosmos client not initialized");

        var database = await _cosmosClient.CreateDatabaseIfNotExistsAsync(_databaseName);
        _container = await database.Database.CreateContainerIfNotExistsAsync(_containerName, "/shortCode");
    }

    public async Task<UrlMapping> CreateUrlAsync(string shortCode, string originalUrl)
    {
        if (_container == null)
            throw new InvalidOperationException("Container not initialized");

        var urlMapping = new CosmosUrlMapping
        {
            Id = Guid.NewGuid().ToString(),
            ShortCode = shortCode,
            OriginalUrl = originalUrl,
            CreatedAt = DateTime.UtcNow
        };

        await _container.CreateItemAsync(urlMapping, new PartitionKey(shortCode));

        return new UrlMapping
        {
            ShortCode = urlMapping.ShortCode,
            OriginalUrl = urlMapping.OriginalUrl,
            CreatedAt = urlMapping.CreatedAt
        };
    }

    public async Task<UrlMapping?> GetUrlAsync(string shortCode)
    {
        if (_container == null)
            throw new InvalidOperationException("Container not initialized");

        try
        {
            var query = new QueryDefinition("SELECT * FROM c WHERE c.shortCode = @shortCode")
                .WithParameter("@shortCode", shortCode);

            var iterator = _container.GetItemQueryIterator<CosmosUrlMapping>(query);
            var results = await iterator.ReadNextAsync();

            var item = results.FirstOrDefault();
            if (item == null) return null;

            return new UrlMapping
            {
                ShortCode = item.ShortCode,
                OriginalUrl = item.OriginalUrl,
                CreatedAt = item.CreatedAt
            };
        }
        catch (CosmosException ex) when (ex.StatusCode == HttpStatusCode.NotFound)
        {
            return null;
        }
    }

    public async Task<List<UrlMapping>> GetAllUrlsAsync()
    {
        if (_container == null)
            throw new InvalidOperationException("Container not initialized");

        var query = new QueryDefinition("SELECT * FROM c ORDER BY c.createdAt DESC");
        var iterator = _container.GetItemQueryIterator<CosmosUrlMapping>(query);

        var urls = new List<UrlMapping>();
        while (iterator.HasMoreResults)
        {
            var results = await iterator.ReadNextAsync();
            urls.AddRange(results.Select(item => new UrlMapping
            {
                ShortCode = item.ShortCode,
                OriginalUrl = item.OriginalUrl,
                CreatedAt = item.CreatedAt
            }));
        }

        return urls;
    }

    public async Task<bool> UrlExistsAsync(string shortCode)
    {
        var url = await GetUrlAsync(shortCode);
        return url != null;
    }

    public async Task CloseAsync()
    {
        if (_cosmosClient != null)
        {
            _cosmosClient.Dispose();
            _cosmosClient = null;
        }
        await Task.CompletedTask;
    }

    private static (string endpoint, string key, string? database) ParseConnectionString(string connectionString)
    {
        var parts = connectionString.Split(';', StringSplitOptions.RemoveEmptyEntries);
        string? endpoint = null, key = null, database = null;

        foreach (var part in parts)
        {
            var keyValue = part.Split('=', 2);
            if (keyValue.Length != 2) continue;

            var propertyName = keyValue[0].Trim();
            var propertyValue = keyValue[1].Trim();

            switch (propertyName)
            {
                case "AccountEndpoint":
                    endpoint = propertyValue;
                    break;
                case "AccountKey":
                    key = propertyValue;
                    break;
                case "Database":
                    database = propertyValue;
                    break;
            }
        }

        if (string.IsNullOrEmpty(endpoint) || string.IsNullOrEmpty(key))
        {
            throw new ArgumentException("Connection string must contain AccountEndpoint and AccountKey");
        }

        return (endpoint, key, database);
    }

    private class CosmosUrlMapping
    {
        public string Id { get; set; } = string.Empty;
        public string ShortCode { get; set; } = string.Empty;
        public string OriginalUrl { get; set; } = string.Empty;
        public DateTime CreatedAt { get; set; }
    }
}
